### PR TITLE
[Group 2] Enable nullable annotations for `Microsoft.Extensions.Options`

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to #43605

Questions:
- Not sure if `failureMessage` should be nullable, but there's never a null-check for it
- In `OptionsFactory.Create` `name` parameter looks like it should be nullable, but `OptionsValidationException` will throw if `name` is `null`

Other questions:
- `dotnet msbuild /t:GenerateReferenceAssemblySource` doesn't update nullable annotations. How can I update `ref` files automatically? (there were a lot of `?`s in this PR)

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs [hunk 1]
```diff
@@ -27,30 +27,30 @@ public ValidateOptions(string name, Func<TOptions, bool> validation, string fail
...
         /// <param name="options">The options instance.</param>
         /// <returns>The <see cref="ValidateOptionsResult"/> result.</returns>
-        public ValidateOptionsResult Validate(string name, TOptions options)
+        public ValidateOptionsResult Validate(string? name, TOptions options)
         {
             // null name is used to configure all named options
             if (Name == null || name == Name)
             {
-                if ((Validation?.Invoke(options)).Value)
+                if ((Validation?.Invoke(options))!.Value)
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs:53]: This doesn't seem safe. If Validation is null, won't this null-ref?

repeated elsewhere

File: src/libraries/Microsoft.Extensions.Options/src/ValidateOptionsResult.cs [hunk 2]
```diff
@@ -39,27 +39,27 @@ public class ValidateOptionsResult
...
         /// Used to describe why validation failed.
         /// </summary>
-        public string FailureMessage { get; protected set; }
+        public string? FailureMessage { get; protected set; }
 
         /// <summary>
         /// Full list of failures (can be multiple).
         /// </summary>
-        public IEnumerable<string> Failures { get; protected set; }
+        public IEnumerable<string?> Failures { get; protected set; } = null!;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Options/src/ValidateOptionsResult.cs:47]: Let's not do this.

```suggestion
        public IEnumerable<string?>? Failures { get; protected set; }
```

File: src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs [hunk 3]
```diff
@@ -10,134 +10,134 @@ public static partial class OptionsServiceCollectionExtensions
...
-        public string Name { get { throw null; } }
-        public virtual void Configure(string name, TOptions options) { }
+        public ConfigureNamedOptions(string? name, System.Action<TOptions>? action) { }
+        public System.Action<TOptions>? Action { get { throw null; } }
+        public string? Name { get { throw null; } }
+        public virtual void Configure(string? name, TOptions options) { }
         public void Configure(TOptions options) { }
     }
-    public partial class ConfigureNamedOptions<TOptions, TDep> : Microsoft.Extensions.Options.IConfigureNamedOptions<TOptions>, Microsoft.Extensions.Options.IConfigureOptions<TOptions> where TOptions : class where TDep : class
+    public partial class ConfigureNamedOptions<TOptions, TDep> : Microsoft.Extensions.Options.IConfigureNamedOptions<TOptions>, Microsoft.Extensions.Options.IConfigureOptions<TOptions> where TOptions : class where TDep : class?
```
[the remark this answers is not in the document]
Comment (HaoK) [reply src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs:35]: Yeah TDep should be non null, as those should be required services for this API to even make sense

File: src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs [hunk 4]
```diff
@@ -204,15 +204,15 @@ public partial class OptionsMonitor<[System.Diagnostics.CodeAnalysis.Dynamically
...
     }
     public partial class OptionsValidationException : System.Exception
     {
-        public OptionsValidationException(string optionsName, System.Type optionsType, System.Collections.Generic.IEnumerable<string> failureMessages) { }
-        public System.Collections.Generic.IEnumerable<string> Failures { get { throw null; } }
+        public OptionsValidationException(string optionsName, System.Type optionsType, System.Collections.Generic.IEnumerable<string?>? failureMessages) { }
+        public System.Collections.Generic.IEnumerable<string?> Failures { get { throw null; } }
         public override string Message { get { throw null; } }
-        public string OptionsName { get { throw null; } }
+        public string? OptionsName { get { throw null; } }
```
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs:215]: Hm, it's not marked as nullable in the `src`, looks like I've misannotated the `ref`. 
Strange that CI still succeeded.

File: src/libraries/Microsoft.Extensions.Options/src/ValidateOptionsResult.cs (hunk 2 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Options/src/ValidateOptionsResult.cs:47]: How about?

```cs
[MemberNotNullWhen(true, nameof(Failures))]
public bool Failed { get; protected set; }
```

`Failures` are always set when fail result is generated (some code would NRE if it isn't)

File: src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs [hunk 5]
```diff
@@ -80,7 +80,7 @@ public TOptions Create(string name)
                 }
                 if (failures.Count > 0)
                 {
-                    throw new OptionsValidationException(name, typeof(TOptions), failures);
+                    throw new OptionsValidationException(name!, typeof(TOptions), failures);
```
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs]: It really seems like it should, bet there's an explicit `throw` if it is. This was one of the questions from the PR description:

> In `OptionsFactory.Create` `name` parameter looks like it should be nullable, but `OptionsValidationException` will throw if `name` is `null`

Since you agree that constructor should accept `null`, I've removed `throw` inside it.

File: src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs [hunk 6]
```diff
@@ -288,74 +288,74 @@ public partial class ValidateOptionsResult
...
-        public string FailureMessage { get { throw null; } protected set { } }
-        public System.Collections.Generic.IEnumerable<string> Failures { get { throw null; } protected set { } }
+        public string? FailureMessage { get { throw null; } protected set { } }
+        public System.Collections.Generic.IEnumerable<string?> Failures { get { throw null; } protected set { } }
         public bool Skipped { get { throw null; } protected set { } }
         public bool Succeeded { get { throw null; } protected set { } }
-        public static Microsoft.Extensions.Options.ValidateOptionsResult Fail(System.Collections.Generic.IEnumerable<string> failures) { throw null; }
-        public static Microsoft.Extensions.Options.ValidateOptionsResult Fail(string failureMessage) { throw null; }
+        public static Microsoft.Extensions.Options.ValidateOptionsResult Fail(System.Collections.Generic.IEnumerable<string?> failures) { throw null; }
+        public static Microsoft.Extensions.Options.ValidateOptionsResult Fail(string? failureMessage) { throw null; }
```
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs]: > Not sure if `failureMessage` should be nullable, but there's never a null-check for it

According to your comments (especially this one), answer the that question is `failureMessage cannot be null`, updated the PR.

File: src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs (hunk 1 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs:53]: Tested it, it will NRE. 
image

I thought since we have `Nullable<t>` struct, it won't actually be `null` and just revert to `default` if there's no value or something. (still don't understand where `Nullable<t>` comes from though)

What should be done here? Can I change it to `if (Validation == null || Validation.Invoke(options))` in this PR?

File: src/libraries/Microsoft.Extensions.Options/src/OptionsValidationException.cs [hunk 7]
```diff
@@ -17,17 +17,17 @@ public class OptionsValidationException : Exception
         /// <param name="optionsName">The name of the options instance that failed.</param>
         /// <param name="optionsType">The options type that failed.</param>
         /// <param name="failureMessages">The validation failure messages.</param>
-        public OptionsValidationException(string optionsName, Type optionsType, IEnumerable<string> failureMessages)
+        public OptionsValidationException(string? optionsName, Type optionsType, IEnumerable<string>? failureMessages)
         {
             Failures = failureMessages ?? new List<string>();
             OptionsType = optionsType ?? throw new ArgumentNullException(nameof(optionsType));
-            OptionsName = optionsName ?? throw new ArgumentNullException(nameof(optionsName));
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsValidationException.cs:24]: Why are we removing the ArgumentNull check here? We should keep it.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsValidationException.cs:24]: Oh, I remember it now, reading the other thread. Sorry.

File: src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs:53]: Maybe the best thing to do here is to make `Validation` be non-nullable, and start throwing in the ctor. Looking at the only callsite of it, I believe that was the original intention:

https://github.com/dotnet/runtime/blob/3c43ef6af7dd0a54e26ae23e4610f4937a7a7e4d/src/libraries/Microsoft.Extensions.Options/src/OptionsBuilder.cs#L368-L373

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs:53]: The other option is to return `Skip` when the `Validation` function is `null`. I don't think we should return `Pass` or `Failed` - we really have no idea.

@HaoK - thoughts here?

File: src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs [hunk 8]
```diff
@@ -163,37 +163,37 @@ public partial class OptionsBuilder<TOptions> where TOptions : class
         public virtual Microsoft.Extensions.Options.OptionsBuilder<TOptions> PostConfigure<TDep1, TDep2, TDep3, TDep4, TDep5>(System.Action<TOptions, TDep1, TDep2, TDep3, TDep4, TDep5> configureOptions) where TDep1 : class where TDep2 : class where TDep3 : class where TDep4 : class where TDep5 : class { throw null; }
         public virtual Microsoft.Extensions.Options.OptionsBuilder<TOptions> Validate(System.Func<TOptions, bool> validation) { throw null; }
         public virtual Microsoft.Extensions.Options.OptionsBuilder<TOptions> Validate(System.Func<TOptions, bool> validation, string failureMessage) { throw null; }
-        public virtual Microsoft.Extensions.Options.OptionsBuilder<TOptions> Validate<TDep>(System.Func<TOptions, TDep, bool> validation) { throw null; }
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs:166]: Why are these changes necessary?

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs:166]: `TDep` is used in `sp.GetRequiredService<TDep>()`, and it has a `notnull` constraint, so `TDep` should also have it

File: src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs (hunk 1 above)
Comment (HaoK) [reply src/libraries/Microsoft.Extensions.Options/src/ValidateOptions.cs]: > make Validation be non-nullable, and start throwing in the ctor

Yeah that was the original intent, there shouldn't be null validation functions

File: src/libraries/Microsoft.Extensions.Options/src/ConfigureNamedOptions.cs [hunk 9]
```diff
@@ -16,7 +16,7 @@ public class ConfigureNamedOptions<TOptions> : IConfigureNamedOptions<TOptions>
         /// </summary>
         /// <param name="name">The name of the options.</param>
         /// <param name="action">The action to register.</param>
-        public ConfigureNamedOptions(string name, Action<TOptions> action)
+        public ConfigureNamedOptions(string? name, Action<TOptions>? action)
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Options/src/ConfigureNamedOptions.cs:19]: I wonder if this makes sense. This is a `ConfigureNamedOptions` class for a "named options". Why would `null` be a valid name?

Even though it doesn't throw ArgNull, I wonder if this should be `string name`.

File: src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs [hunk 10]
```diff
@@ -48,7 +48,7 @@ public OptionsFactory(IEnumerable<IConfigureOptions<TOptions>> setups, IEnumerab
         /// <summary>
         /// Returns a configured <typeparamref name="TOptions"/> instance with the given <paramref name="name"/>.
         /// </summary>
-        public TOptions Create(string name)
+        public TOptions Create(string? name)
```
[the remark this answers is not in the document]
Comment (HaoK) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs:51]: Right I think null name is normalized to Options.DefaultName generally, I think null name is only used to signal configuring all option instances in some places if I remember correctly

File: src/libraries/Microsoft.Extensions.Options/src/ConfigureNamedOptions.cs (hunk 9 above)
Comment (HaoK) [reply src/libraries/Microsoft.Extensions.Options/src/ConfigureNamedOptions.cs:19]: It avoided needing to introduce a special value for ConfigureAll, because null name is generally illegal, so it was safe to use for https://github.com/dotnet/runtime/blob/22a71c8a1d4ff224e739d49eaf75597b5a89b8fb/src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs#L67

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/src/ConfigureNamedOptions.cs:19]: Ah this makes sense. Thanks!

File: src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs:51]: I believe this one is still valid. What would "Create `all`" mean?

If you want to create an Options, you have to pass in a non-null string. And if the Options doesn't have a name, you use `Options.DefaultName` (which is `string.Empty`).

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs:51]: You can call nullable APIs from a method that takes in a non-nullable value. I don't see that as a reason the decide what `Create` should be allowed to take.

`null` names means "configure all named options". But the `OptionsFactory.Create` API isn't dealing with "all" options. It is dealing with _this one option_ - the thing it is creating. And saying "to create an option, you need to give it a name or use `Options.DefaultName`" makes sense to me.

[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs]: Yeah, but it's strange that _all_ methods that are called from `Create` accept `null` (not counting `OptionsValidationException` since there's no business logic in there), and there's no much in logic in `Create` itself outside calling those methods. I guess, by the logic you described `CreateInstance` also shouldn't accept `null`.

Marked both `Create` and `CreateInstance`'es `name` parameter as non-nullable.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs]: >  by the logic you described CreateInstance also shouldn't accept null.

Correct. That was my thinking as well.

**Review by eerhardt — APPROVED** [review]
Thanks again for all your hard work here @maxkoshevoi!

LGTM
